File: Trees/Next Right Node/Next_Right_Node.cpp
//Problem Name: Next Right Node
//Problem Link: https://www.geeksforgeeks.org/problems/next-right-node/1 

class Solution {
  public:
    Node* nextRightNode(Node *root, int key) {
        Node* ans = new Node(-1);
        if(root == NULL) return ans;
        queue<Node*> q;
        q.push(root);
        while(!q.empty()) {
            int sz = q.size();
            vector<Node*> level;
            for(int i = 0;i<sz;i++) {
                Node* node = q.front();
                q.pop();
                if(node->left!=NULL) q.push(node->left);
                if(node->right!=NULL) q.push(node->right);
                level.push_back(node);
            }
            int ind = -1;
            for(int i = 0;i<level.size();i++) {
                Node* temp = level[i];
                if(temp->data == key) {
                    ind = i;
                    break;
                }
            }
            if(ind>-1 && ind+1<level.size()) {
                ans = level[ind+1];
                break;
            }
        }
        return ans;
    }
    Node *nextRight(Node *root, int key) {
        // code here
        return nextRightNode(root, key);
    }
};

File: Trees/Next Right Node/Next_Right_Node.java
//Problem Name: Next Right Node
//Problem Link: https://www.geeksforgeeks.org/problems/next-right-node/1 

class Solution {
    Node nextRightNode(Node root, int key) {
        Node ans = new Node(-1);
        if(root == null) return ans;
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while(!q.isEmpty()){
            int sz = q.size();
            List<Node> level = new ArrayList<>();
            for(int i = 0;i<sz;i++) {
                Node node = q.poll();
                if(node.left!=null) q.add(node.left);
                if(node.right!=null) q.add(node.right);
                level.add(node);
            }
            Node[] newLevel = level.stream().toArray(Node[]::new);
            int ind = -1;
            for(int i = 0;i<newLevel.length;i++){
                if(newLevel[i].data == key) {
                    ind = i;
                    break;
                }
            }
            if(ind>-1 && ind+1<newLevel.length) {
                ans = newLevel[ind+1];
                break;
            }
        }
        return ans;
    }
    Node nextRight(Node root, int key) {
        return nextRightNode(root, key);
    }
}